Session recording: persistent URLs, session_id, UI changes

macobo opened the pull request:
### Changes

- We now don't show $snapshot events in various views
- We hide system info on snapshot events
- New querying system for loading data from API
- Add sharable/persistent URLs for session recording
- Use Drawer over Modal to be able to fit more things into it
- Add Next/Previous buttons to session recording player

Cypress tests are upcoming in follow-ups.

![Screenshot from 2020-10-21 15-39-14](https://user-images.githubusercontent.com/148820/96720573-9aac7800-13b3-11eb-992c-07fa5b80af64.png)

Some of the BE and Kea-related code feels sort of disgusting - any and all pointers appreciated!

### Commits

﻿- Hide $session_id properties
- Exclude $snapshot events from event views
- Add end_time to sessions queries
- Add WIP way to return session recording
- Query for feature flags in a consistent way in frontend
- Improve documentation on featureFlagLogic
- Update frontend types
- Set up endpoint for querying event data for rrweb
- Sort snapshot data in BE before sending to FE
- Show sessions player in a drawer, load data from API
- Show loader while replay is loading
- Make player full-width
- Remove debug code
- Implement persistent URLs for sessions
- Use only $session_id for filtering sessions
- WIP
- Add `array.flatMap` to tsconfig
- Add selectors for prev/next session recording
- Clean up whitespace
- Make next and previous buttons work
- Patch prev/next behavior when not visible in page
- Session recording query testing


## Checklist

- [x] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [x] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

macobo committed:
Hide $session_id properties

macobo committed:
Exclude $snapshot events from event views

macobo committed:
Add end_time to sessions queries

This might be used by SessionRecording queries

macobo committed:
Add WIP way to return session recording

Note that this mutates existing responses - might not be the best way to
do this.

macobo committed:
Query for feature flags in a consistent way in frontend

macobo committed:
Improve documentation on featureFlagLogic

macobo committed:
Update frontend types

macobo committed:
Set up endpoint for querying event data for rrweb

macobo committed:
Show sessions player in a drawer, load data from API

This makes for a smoother experience (IMO, will iterate). At the very
least, drawers enable us to have differing urls via kea

macobo committed:
Sort snapshot data in BE before sending to FE

macobo committed:
Show loader while replay is loading

macobo committed:
Make player full-width

macobo committed:
Remove debug code

macobo committed:
Implement persistent URLs for sessions

This should make it possible to share recordings with team.

The URLs are currently quite long - perhaps we can get rid of
distinct_id long-term within it?

macobo committed:
WIP

macobo committed:
Use only $session_id for filtering sessions

There's a small risk of collisions, but this cleans code up a lot

macobo committed:
Add selectors for prev/next session recording

macobo committed:
Clean up whitespace

macobo committed:
Add `array.flatMap` to tsconfig

macobo committed:
Make next and previous buttons work

macobo committed:
Patch prev/next behavior when not visible in page

macobo committed:
Session recording query testing

macobo commented on frontend/src/scenes/sessions/sessionsTableLogic.ts:
```diff
@@ -109,19 +167,27 @@ export const sessionsTableLogic = kea<sessionsTableLogicType<Moment, SessionType
-        },
+        setFilters: () => buildURL(values.selectedDateURLparam, values.sessionRecordingId),
+        loadSessionPlayer: () => buildURL(values.selectedDateURLparam, values.sessionRecordingId),
+        closeSessionPlayer: () => buildURL(values.selectedDateURLparam, values.sessionRecordingId),
```
This + urlToAction feels like it violates DRY a lot, plus this logic is getting huge.

Is there a pattern here that would help clean things up?

macobo commented on posthog/api/event.py:
```diff
@@ -311,3 +312,17 @@ def sessions(self, request: request.Request) -> response.Response:
+    # - session_recording_id: (string) id of the session recording
+    # ******************************************
+    @action(methods=["GET"], detail=False)
+    def session_recording(self, request: request.Request, *args: Any, **kwargs: Any) -> response.Response:
```
Not sure where this endpoint belongs/how we're structuring our api, thoughts?

macobo commented on tsconfig.json:
```diff
@@ -26,7 +26,7 @@
         "incremental": true, // Enable incremental compilation by reading/writing information from prior compilations to a file on disk
         "noFallthroughCasesInSwitch": true, // Report errors for fallthrough cases in switch statement
-        "lib": ["dom"]
+        "lib": ["dom", "es2019.array"]
```
Needed this to get access to `Array.flatMap` - any reason not to enable this? cc @mariusandra 

https://caniuse.com/?search=array.flatmap

mariusandra replied:
We actually had someone use a derpy browser before, which didn't support [Object.entries](https://caniuse.com/?search=object.entries), prompting the addition [of this](
https://github.com/PostHog/posthog/blob/43816096d0ecafc2a9370d976f806c8a90835699/frontend/src/index.html#L9) to index.html

Let's add flatMap there as well for now?

macobo committed:
Fix typing errors

macobo committed:
Make Sessions queries work properly

It was previously overwriting things even in queries not returning a
list of sessions

macobo committed:
Blacklist browsers w/o flatMap (PR review)

mariusandra reviewed (commented):
I can't for the life of me get the feature flag enabled in localhost to test this. I'm going to send the few pieces of feedback I had and will try to run it later.

mariusandra commented:
Some UX feedback, again not sure if this goes under the scope of this PR:

- There's no feedback that this was actually saved:
![2020-10-23 12 29 53](https://user-images.githubusercontent.com/53387/96993354-7e334b80-152b-11eb-8ab5-3a851ba698e4.gif)
... I have to reload the page like it's 1999 to make sure of that.

- Looking at the playback, I immediately get some questions, such as:
  - ![2020-10-23 12 32 59](https://user-images.githubusercontent.com/53387/96993697-f0a42b80-152b-11eb-8c53-016520b08c4e.gif)
  - What page (URL) is my visitor on?
  - At what time/date was the session 
  - The playback progress bar + buttons could be just on row if there's enough space (E.g. not on a phone)
  - Some alignment issue with the "skip inactive" switch and label

mariusandra commented:
One more, the back button doesn't work:

![2020-10-23 12 38 09](https://user-images.githubusercontent.com/53387/96994382-f0f0f680-152c-11eb-9522-3f4e5f0ed808.gif)

I guess it's a quick fix inside the `'/person/*'` `urlToAction` key...

mariusandra reviewed (changes requested):
No strong comments here re: the session_id part, just a few small things on the periphery to improve.

macobo commented on frontend/src/scenes/sessions/SessionsPlayerButton.tsx:
```diff
@@ -1,39 +1,29 @@
+                        event.stopPropagation()
+                        loadSessionPlayer(sessionRecordingId)
+                    }}
+                ></PlayCircleOutlined>
```
Will fix, I think the eslint rule is a good idea :+1:

macobo commented on posthog/api/event.py:
```diff
@@ -107,7 +108,7 @@ def to_representation(self, instance):
 class EventViewSet(viewsets.ModelViewSet):
     renderer_classes = tuple(api_settings.DEFAULT_RENDERER_CLASSES) + (csvrenderers.PaginatedCSVRenderer,)
-    queryset = Event.objects.all()
+    queryset = Event.objects.all().exclude(event="$snapshot")
```
I don't know the context behind excluding $feature_flag_called - isn't it a event like others?

The fact I'm excluding $snapshot here is indicating I'm actually abusing Events a bit here to do a different thing.

mariusandra replied:
It's just a lot of noise. People have complained that it makes the "all events" list unusable and if you search for `$feature_flag_called` in the codebase, you'll just find a bunch of lines like `and event != "$feature_flag_called"`

macobo commented on posthog/queries/session_recording.py:
```diff
@@ -0,0 +1,41 @@
+from typing import Any, Dict, List
```
Yeah - I don't know (yet) how to make this stuff with clickhouse, this is why I've been pushing it in standups. :)

Will get to it as I wrap up the pg storyline.

macobo commented on posthog/queries/session_recording.py:
```diff
@@ -0,0 +1,41 @@
+class SessionRecording(BaseQuery):
+    def run(self, filter: Filter, team: Team, *args, **kwargs) -> List[Dict[str, Any]]:
+        events = Event.objects.filter(team=team, event="$snapshot").filter(
+            **{"properties__$session_id": kwargs["session_recording_id"]}
```
:+1: On my list, will be done in a follow-up.

macobo commented:
> I can't for the life of me get the feature flag enabled in localhost to test this.

Yikes. I think this is us feeling the pain of our customers as well - #1996

> Looking at the playback, I immediately get some questions, such as:

Agreed. There is a lot we could expose in this UI - part of the reason for using Drawer over modal is to give us the tools to do that. Noted these things down + will brainstorm a few more, figure out the UX and add them in future PRs.

macobo commented:
Re feedback when saving: Agreed, however I'll be redoing the whole design on settings view in future PRs.

Toggles not giving feedback is a general thing - the pattern came from another setup page toggle currently.

macobo committed:
Close sessions player on back

macobo commented:
Updated the back button behavior, but also noticing:

1. If you close the drawer, whole sessions page data gets loaded again
2. If you press back, it might jump between URLs real quick due to the sequence of actions being triggered.

It think both have to do with the kea issue you raised and I suggest tackling them separately from this PR (it's blocking making follow-up changes).

mariusandra committed:
fix back button and small bugs

mariusandra commented:
I tried my best at fixing the back button. This logic with the urls and actions is a bit of a mess and should be cleared up, yet now it works (on both session and person pages)... so I'd consider this good enough.

I saw a few other TS issues that needed fixing that got done in a flyby.

mariusandra commented on frontend/src/scenes/sessions/SessionsPlayer.tsx:
```diff
@@ -7,7 +7,7 @@ export default function SessionsPlayer({ events }: { events: eventWithTime[] }):
 
     useEffect(() => {
-        if (target.current) {
+        if (target.current && events) {
```
@macobo the only thing I'm not 100% convinced of is this `&& events` part. Without it the back/forward buttons would crash, as rrwebPlayer got initialized with some empty data when going "forward" from /sessions to /sessions?id=..

Seeing how this `events` is not in the `[]` deps array I'm not sure if what I'm doing is legit or not.

mariusandra reviewed (approved):
fine to merge from my side, yet I have one comment regarding a fix I did.
